### PR TITLE
[Issue #3985] closing opportunities email

### LINKED ISSUE #3985: Deploy API gateway
### Summary

\

### Acceptance criteria

\

### PR DESCRIPTION
**Pull request by mikehgrantsgov**
## Summary
Fixes #3985

## Changes proposed
Add a new table `UserOpportunityNotificationLog` to store records of users/opportunities we have notified on when closing.
Add logic to `generate_notifications.py` to collect and send notifications for closing opportunities.
Add tests to cover functionality (sending and avoiding duplicate sending)

## Context for reviewers
Users should receive notifications within 2 weeks of a closing opportunity that they have bookmarked. 

## Additional information
Email content body does not yet fully conform to content doc, but putting up a draft PR to confirm approach.

### CODE REVIEW

File: api/tests/conftest.py [hunk 1]
```diff
@@ -445,6 +446,10 @@ def truncate_opportunities(self, db_session):
         class implementation.
         """
 
+        notification_logs = db_session.query(UserOpportunityNotificationLog).all()
```
Comment (mikehgrantsgov) [inline api/tests/conftest.py:449]: @chouinar should this be handled by a cascading delete? Not sure if we want to keep these records around in the event a user or opportunity is ever deleted, so I didn't add a cascade delete.

File: api/src/db/models/user_models.py [hunk 2]
```diff
@@ -144,3 +144,19 @@ class UserNotificationLog(ApiSchemaTable, TimestampMixin):
...
+        UUID, primary_key=True, default=uuid.uuid4
+    )
+
+    user_id: Mapped[uuid.UUID] = mapped_column(ForeignKey(User.user_id), index=True)
+    user: Mapped[User] = relationship(User)
+
+    opportunity_id: Mapped[int | None] = mapped_column(
+        BigInteger, ForeignKey(Opportunity.opportunity_id), index=True, nullable=True
+    )
+    opportunity: Mapped[Opportunity] = relationship("Opportunity")
```
Comment (chouinar) [inline api/src/db/models/user_models.py:162]: When would we want opportunity_id to be null?

File: api/src/task/notifications/generate_notifications.py [hunk 3]
```diff
@@ -48,6 +50,7 @@ def run_notification_task(db_session: db.Session, search_client: search.SearchCl
 class NotificationConstants:
     OPPORTUNITY_UPDATES = "opportunity_updates"
     SEARCH_UPDATES = "search_updates"
+    CLOSING_DATE_REMINDER = "closing_date_reminder"  #
```
Comment (chouinar) [inline api/src/task/notifications/generate_notifications.py:53]:
```suggestion
    CLOSING_DATE_REMINDER = "closing_date_reminder"
```

File: api/tests/conftest.py (hunk 1 above)
Comment (chouinar) [reply api/tests/conftest.py:449]: Because we have a foreign key from the notification log table to the opportunity table, we need a cascading delete or special logic elsewhere to do something in this case (null out the column? Might be more confusing to keep it).

File: infra/api/app-config/env-config/environment-variables.tf [hunk 4]
```diff
@@ -62,5 +62,10 @@ locals {
       manage_method     = "manual"
       secret_store_name = "/api/${var.environment}/frontend-login-redirect-url"
     }
+
+    FRONTEND_BASE_URL = {
+      manage_method     = "manual"
+      secret_store_name = "/api/${var.environment}/frontend-base-url"
+    }
```
[the remark this answers is not in the document]
Comment (mikehgrantsgov) [reply infra/api/app-config/env-config/environment-variables.tf:69]: These have been created for dev, staging, and prod at the per-environment paths.

**Review by chouinar — APPROVED** [review]
LGTM!
